The job **`dist-i586-gnu-i586-i686-musl`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/actions/runs/3504681361/jobs/5870516754) [(plain)](https://github.com/rust-lang-ci/rust/commit/d9c75547237506c7038686018184bf9554859730/checks/9593841128/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
failures:

---- [codegen] src/test/codegen/mem-replace-direct-memcpy.rs stdout ----

error: verification with 'FileCheck' failed
status: exit status: 1
command: "/checkout/obj/build/x86_64-unknown-linux-gnu/llvm/build/bin/FileCheck" "--input-file" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/mem-replace-direct-memcpy/mem-replace-direct-memcpy.ll" "/checkout/src/test/codegen/mem-replace-direct-memcpy.rs" "--allow-unused-prefixes" "--check-prefixes" "CHECK,NONMSVC"
stdout: none
--- stderr -------------------------------
/checkout/src/test/codegen/mem-replace-direct-memcpy.rs:21:11: error: CHECK: expected string not found in input
// CHECK: call void @llvm.memcpy.{{.+}}({{i8\*|ptr}} align 1 %{{.*}}, {{i8\*|ptr}} align 1 %src, i{{.*}} 1, i1 false)
          ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/mem-replace-direct-memcpy/mem-replace-direct-memcpy.ll:6:21: note: scanning from here
; core::mem::replace
                    ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/mem-replace-direct-memcpy/mem-replace-direct-memcpy.ll:22:2: note: possible intended match here
 call void @llvm.memcpy.p0.p0.i32(ptr align 1 %dest, ptr align 1 %src1, i32 1, i1 false)

Input file: /checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/mem-replace-direct-memcpy/mem-replace-direct-memcpy.ll
Check file: /checkout/src/test/codegen/mem-replace-direct-memcpy.rs


-dump-input=help explains the following input dump.
Input was:
<<<<<<
<<<<<<
            1: ; ModuleID = 'mem_replace_direct_memcpy.6e601deb-cgu.0' 
            2: source_filename = "mem_replace_direct_memcpy.6e601deb-cgu.0" 
            3: target datalayout = "e-m:e-p:32:32-p270:32:32-p271:32:32-p272:64:64-f64:32:64-f80:32-n8:16:32-S128" 
            4: target triple = "i586-unknown-linux-gnu" 
            6: ; core::mem::replace 
            6: ; core::mem::replace 
check:21'0                         X error: no match found
            7: ; Function Attrs: inlinehint nonlazybind uwtable 
check:21'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
            8: define internal i8 @_ZN4core3mem7replace17h8eaa2e5bbcc07fd8E(ptr noalias noundef align 1 dereferenceable(1) %dest, i8 %src) unnamed_addr #0 personality ptr @rust_eh_personality { 
check:21'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
            9: start: 
check:21'0     ~~~~~~~
           10:  %0 = alloca { ptr, i32 }, align 4 
check:21'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           11:  %tmp = alloca i8, align 1 
check:21'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~
            .
            .
           17:  
           17:  
check:21'0     ~
           18: bb4: ; preds = %start 
check:21'0     ~~~~~~~~~~~~~~~~~~~~~~
           19:  call void @llvm.lifetime.end.p0(i64 1, ptr %tmp) 
check:21'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           20:  call void @llvm.lifetime.start.p0(i64 1, ptr %src1) 
check:21'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           21:  store i8 %src, ptr %src1, align 1 
check:21'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           22:  call void @llvm.memcpy.p0.p0.i32(ptr align 1 %dest, ptr align 1 %src1, i32 1, i1 false) 
check:21'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
check:21'1      ?                                                                                        possible intended match
           23:  call void @llvm.lifetime.end.p0(i64 1, ptr %src1) 
check:21'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           24:  ret i8 %self 
check:21'0     ~~~~~~~~~~~~~~
           25:  
check:21'0     ~
           26: bb3: ; No predecessors! 
check:21'0     ~~~~~~~~~~~~~~~~~~~~~~~~
           27:  br i1 true, label %bb2, label %bb1 
check:21'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
            .
            .
>>>>>>
------------------------------------------
```

</details>

--- a/compiler/rustc_ty_utils/src/layout.rs
+++ b/compiler/rustc_ty_utils/src/layout.rs
@@ -140,8 +140,13 @@ fn univariant_uninterned<'tcx>(
         let optimizing = &mut inverse_memory_index[..end];
         let effective_field_align = |f: &TyAndLayout<'_>| {
             if let Some(pack) = pack {
+                // return the packed alignment in bytes
                 f.align.abi.min(pack).bytes()
             } else {
+                // returns log2(effective-align).
+                // This is ok since `pack` applies to all fields equally.
+                // The calculation assumes that size is an integer multiple of align, except for ZSTs.
+                //
                 // group [u8; 4] with align-4 or [u8; 6] with align-2 fields
                 f.align.abi.bytes().max(f.size.bytes()).trailing_zeros() as u64
             }
@@ -165,15 +170,23 @@ fn univariant_uninterned<'tcx>(
                     optimizing.sort_by_key(|&x| {
                         // Place ZSTs first to avoid "interesting offsets",
                         // especially with only one or two non-ZST fields.
+                        // Then place largest alignments first, largest niches within an alignment group last
                         let f = &fields[x as usize];
-                        (!f.is_zst(), cmp::Reverse(effective_field_align(f)))
+                        let niche_size = f.largest_niche.map_or(0, |n| n.available(cx));
+                        (!f.is_zst(), cmp::Reverse(effective_field_align(f)), niche_size)
                     });
                 }
 
                 StructKind::Prefixed(..) => {
                     // Sort in ascending alignment so that the layout stays optimal
-                    // regardless of the prefix
-                    optimizing.sort_by_key(|&x| effective_field_align(&fields[x as usize]));
+                    // regardless of the prefix.
+                    // And put the largest niche in an alignment group at the end
+                    // so it can be used as discriminant in jagged enums
+                    optimizing.sort_by_key(|&x| {
+                        let f = &fields[x as usize];
+                        let niche_size = f.largest_niche.map_or(0, |n| n.available(cx));
+                        (effective_field_align(f), niche_size)
+                    });
                 }
             }
 

--- a/compiler/rustc_ty_utils/src/layout.rs
+++ b/compiler/rustc_ty_utils/src/layout.rs
@@ -140,14 +140,10 @@ fn univariant_uninterned<'tcx>(
         let optimizing = &mut inverse_memory_index[..end];
         let effective_field_align = |f: &TyAndLayout<'_>| {
             if let Some(pack) = pack {
-                f.align.abi.min(pack)
-            } else if f.size.bytes().is_power_of_two() && f.size.bytes() >= f.align.abi.bytes() {
-                // Try to put fields which have a 2^n size and smaller alignment together with
-                // fields that have an alignment matching that size.
-                // E.g. group [u8; 4] with u32 fields
-                Align::from_bytes(f.align.abi.bytes()).unwrap_or(f.align.abi)
+                f.align.abi.min(pack).bytes()
             } else {
-                f.align.abi
+                // group [u8; 4] with align-4 or [u8; 6] with align-2 fields
+                f.align.abi.bytes().max(f.size.bytes()).trailing_zeros() as u64
             }
         };
 

--- a/compiler/rustc_ty_utils/src/layout.rs
+++ b/compiler/rustc_ty_utils/src/layout.rs
@@ -138,8 +138,17 @@ fn univariant_uninterned<'tcx>(
     if optimize {
         let end = if let StructKind::MaybeUnsized = kind { fields.len() - 1 } else { fields.len() };
         let optimizing = &mut inverse_memory_index[..end];
-        let field_align = |f: &TyAndLayout<'_>| {
-            if let Some(pack) = pack { f.align.abi.min(pack) } else { f.align.abi }
+        let effective_field_align = |f: &TyAndLayout<'_>| {
+            if let Some(pack) = pack {
+                f.align.abi.min(pack)
+            } else if f.size.bytes().is_power_of_two() && f.size.bytes() >= f.align.abi.bytes() {
+                // Try to put fields which have a 2^n size and smaller alignment together with
+                // fields that have an alignment matching that size.
+                // E.g. group [u8; 4] with u32 fields
+                Align::from_bytes(f.align.abi.bytes()).unwrap_or(f.align.abi)
+            } else {
+                f.align.abi
+            }
         };
 
         // If `-Z randomize-layout` was enabled for the type definition we can shuffle
@@ -161,14 +170,14 @@ fn univariant_uninterned<'tcx>(
                         // Place ZSTs first to avoid "interesting offsets",
                         // especially with only one or two non-ZST fields.
                         let f = &fields[x as usize];
-                        (!f.is_zst(), cmp::Reverse(field_align(f)))
+                        (!f.is_zst(), cmp::Reverse(effective_field_align(f)))
                     });
                 }
 
                 StructKind::Prefixed(..) => {
                     // Sort in ascending alignment so that the layout stays optimal
                     // regardless of the prefix
-                    optimizing.sort_by_key(|&x| field_align(&fields[x as usize]));
+                    optimizing.sort_by_key(|&x| effective_field_align(&fields[x as usize]));
                 }
             }
 

